SAN/code/network_simulations/benchmark_timing.jl

```diff
@@ -0,0 +1,256 @@
+unique!(data) # delete duplicate rows, use `nonunique(data)` to see if there are any duplicates
+data = data[isequal.(data.qt_dt,195), :] # keep quarter == 195 = 2008q4
+sort!(data, :assets, rev = true)
+data = data[1:N,:] # keep small number of nodes, for testing
+#N = size(data,1) # number of nodes
+units = 1e6;
+data[:,[:w, :c, :assets, :p_bar, :b]] .= data[!,[:w, :c, :assets, :p_bar, :b]]./units
+# data.b[:] .= missing
+# data.c[:] .= missing
+
+col_with_miss = names(data)[[any(ismissing.(col)) for col = eachcol(data)]] # columns with at least one missing
+data_nm = coalesce.(data, data.assets/1.5) # replace missing by a value
+nm_c = findall(x->x==0,ismissing.(data.c))
+nm_b = findall(x->x==0,ismissing.(data.b))
+dropmissing(data, [:delta, :delta_alt, :w, :assets, :p_bar]) # remove type missing
+
+names(data) # column names
+describe(data)
+show(data, allcols = true)
+
+p_bar = convert(Array{Float64},data.p_bar)
+assets = convert(Array{Float64},data.assets)
+delta = convert(Array{Float64},data.delta)
+w=convert(Array{Float64},data.w)
+g0 = 0.05   # bankruptcy cost
+
+# initial guess
+rng =  Random.seed!(123)
+A0 = zeros(N,N)
+c0 =  data_nm.c 
+b0 = data_nm.b
+α0 = 1.0*ones(N)
+β0= []
+for i=1:N
+    function ff!(F, x)
+        F[1] = 1-cdf(Beta(α0[i],x[1]),w[i]/c0[i])-delta[i] 
+    end
+    if i==1
+        sol = nlsolve(ff!,[50.0])
+    else
+        sol = nlsolve(ff!,[1.0])
+    end
+    @test converged(sol)
+    push!(β0,sol.zero[1])
+end
+
+dist_pdf(x,p) = exp(DistributionsAD.logpdf(DistributionsAD.Beta(p[1],p[2]),x[1]))
+
+function dist_cdf(p...)
+    α,β,wc = p[1], p[2], p[3]
+    s(x,wc) = 1 ./(1 .+exp.(-1000*(x-wc))) # approximates indicator function of x>=wc
+    obj(x,p) = dist_pdf(x,p)*s(x,wc...)
+    prob = QuadratureProblem(obj,0.0,1.0,[α...,β...])
+    Quadrature.solve(prob,QuadGKJL(),reltol=1e-5,abstol=1e-5)[1]
+end
+
+function dist_pdf_mv(x,params)
+ # α,β,A,b,c = params[1][1:N], params[1][N+1:2*N], params[1][2*N+1:2*N+N^2], params[1][2*N+N^2+1:3*N+N^2], params[1][3*N+N^2+1:4*N+N^2]
+     α,β,A,b,c = params[1:N], params[N+1:2*N], params[2*N+1:2*N+N^2], params[2*N+N^2+1:3*N+N^2], params[3*N+N^2+1:4*N+N^2]
+     A = reshape([A...],N,N) 
+     joint_pdf = 1.0;
+     for i=1:N
+         joint_pdf = joint_pdf*dist_pdf(x[i],[α[i],β[i]]) 
+     end
+ 
+    function contraction(x,p)
+        min.(p_bar, max.((1 .+g0).*(A'*p .+ c .- x.*c) .- g0.*p_bar,0))
+    end
+    contraction_iter(x, n::Integer) = n <= 0 ? p_bar  : contraction(x,contraction_iter(x,n-1))
+    loss_x(x) = -sum(contraction_iter(x,2)).*dist_pdf(x,[α...,β...])   
+    loss_x(x)*joint_pdf
+end
+
+free_mem = []
+
+function ev(params...)
+    prob = QuadratureProblem(dist_pdf_mv,zeros(N),ones(N),[params...])
+    global free_mem = [free_mem; Sys.free_memory()/1000/2^20]
```

Comment:
I am also still unsure about the units for memory. Here's a CloudWatch capture and the content of the .csv file that the code outputs. Does the  memory usage differ by an order of magnitude? Or am I not reading things well? I've made that mistake before! But can't square these two outputs. Could you take a look and see if you can?

### CloudWatch
![image](https://user-images.githubusercontent.com/66431383/112920087-43cf6b00-90d6-11eb-94e7-c452c35b1ac2.png)

### meta_data_3.csv
```
nodes,total_time_min,total_mem_gb,max_mem,error,tests_passed,total_tests
3,7.504939448833466,64.88249125099182,0.19004687500000017,ITERATION_LIMIT,5,9
```

Reply:
My way of determining max memory may be misleading. I basically record the memory whenever a particular function is called and so that may not be reflecting of the actual max memory usage overall. I found very little while searching that suggest there is a good way to record the max memory a function uses except monitoring like in cloudwatch.

Reply:
Even if it doesn't capture max mem, I think what you did is good and informative, and we should keep it in. It certainly captures mem usage at that particular point in the code, which is less than or equal to max mem used throughout.

What puzzles me is that whatever is being measured inside julia gives 64Gb. If it is true that 64gb is less than or equal to max mem, then it should be smaller than the 13Gb that CloudWatch shows as max mem used.